rmf_site_format/src/workcell.rs

```diff
@@ -99,23 +107,115 @@ pub struct Inertial {
     pub inertia: Inertia,
 }
 
-#[derive(Serialize, Deserialize, Debug, Default, Clone)]
-#[cfg_attr(feature = "bevy", derive(Bundle))]
-pub struct Link {
-    pub name: NameInWorkcell,
-    pub inertial: Inertial,
-    #[serde(skip)]
-    pub marker: LinkMarker,
+#[derive(Serialize, Deserialize, Debug, Clone, Copy, PartialEq)]
+#[cfg_attr(feature = "bevy", derive(Component))]
+pub enum JointType {
+    Fixed,
+    Revolute,
+    Prismatic,
+    Continuous,
+}
+
+impl JointType {
+    pub fn label(&self) -> String {
+        match &self {
+            JointType::Fixed => "Fixed",
+            JointType::Revolute => "Revolute",
+            JointType::Prismatic => "Prismatic",
+            JointType::Continuous => "Continuous",
+        }
+        .to_string()
+    }
 }
 
-#[derive(Debug, Default, Clone)]
+impl From<&JointType> for urdf_rs::JointType {
+    fn from(joint_type: &JointType) -> Self {
+        match joint_type {
+            JointType::Fixed => urdf_rs::JointType::Fixed,
+            JointType::Revolute => urdf_rs::JointType::Revolute,
+            JointType::Prismatic => urdf_rs::JointType::Prismatic,
+            JointType::Continuous => urdf_rs::JointType::Continuous,
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Default, Clone)]
 #[cfg_attr(feature = "bevy", derive(Component))]
-pub struct LinkMarker;
+pub struct JointLimit {
+    pub lower: f32,
+    pub upper: f32,
+    pub effort: f32,
+    pub velocity: f32,
+}
+
+impl From<&urdf_rs::JointLimit> for JointLimit {
+    fn from(limit: &urdf_rs::JointLimit) -> Self {
+        Self {
+            lower: limit.lower as f32,
+            upper: limit.upper as f32,
+            effort: limit.effort as f32,
+            velocity: limit.velocity as f32,
+        }
+    }
+}
+
+impl From<&JointLimit> for urdf_rs::JointLimit {
+    fn from(limit: &JointLimit) -> Self {
+        Self {
+            lower: limit.lower as f64,
+            upper: limit.upper as f64,
+            effort: limit.effort as f64,
+            velocity: limit.velocity as f64,
+        }
+    }
+}
 
 #[derive(Serialize, Deserialize, Debug, Default, Clone)]
-#[cfg_attr(feature = "bevy", derive(Bundle))]
+#[cfg_attr(feature = "bevy", derive(Component))]
+pub struct JointAxis([f32; 3]);
+
+impl From<&urdf_rs::Axis> for JointAxis {
+    fn from(axis: &urdf_rs::Axis) -> Self {
+        Self(axis.xyz.map(|t| t as f32))
+    }
+}
+
+impl From<&JointAxis> for urdf_rs::Axis {
+    fn from(axis: &JointAxis) -> Self {
+        Self {
+            xyz: urdf_rs::Vec3(axis.0.map(|v| v as f64)),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
 pub struct Joint {
     pub name: NameInWorkcell,
+    pub joint_type: JointType,
```

Comment:
I think we're missing a huge opportunity here to take advantage of Rust enums.

Every different joint type is potentially going to have very different property fields. For single-dimensional joint types like prismatic and revolute you can mostly reduce the properties down to a single set of joint limits and a single axis, but other joint types, such as a [universal joint](https://en.wikipedia.org/wiki/Universal_joint) may require you to define two or more axes, and multiple dimensions of joint limits. There may be arbitrarily exotic joints that we want to support in the future.

So I'd recommend we do something more like this:

```
pub struct Joint {
    pub name: NameInWorkcell,
    pub properties: JointProperties,
}

pub enum JointProperties {
    Fixed,
    Prismatic(SingleDofJoint),
    Revolute(SingleDofJoint),
    Continuous(SingleDofJoint),
}

pub struct SingleDofJoint {
    pub limit: Option<JointLimit>,
    pub axis: JointAxis, // This should be required, there's no reasonable default
}
```

Then we can add more variants at any time without disrupting the existing single-DoF variants.

Reply:
https://github.com/open-rmf/rmf_site/pull/177/commits/53357a6cf5cdf9ec2cb9327d11f0437b104325ac. There is now a new `RangeLimits` enum to contain the different types of limits